src/workerd/api/encoding.c++

```diff
@@ -460,29 +461,204 @@ kj::Maybe<jsg::JsString> TextDecoder::decodePtr(
+    *out = static_cast<char>(c);
+    return 1;
+  } else if (c < 0x800) {
+    out[0] = static_cast<char>(0xC0 | (c >> 6));
+    out[1] = static_cast<char>(0x80 | (c & 0x3F));
+    return 2;
+  } else {
+    out[0] = static_cast<char>(0xE0 | (c >> 12));
+    out[1] = static_cast<char>(0x80 | ((c >> 6) & 0x3F));
+    out[2] = static_cast<char>(0x80 | (c & 0x3F));
+    return 3;
+  }
+}
+
+// Encode a valid surrogate pair to UTF-8
+inline void encodeSurrogatePair(char16_t lead, char16_t trail, char* out) {
+  uint32_t codepoint = 0x10000 + (((lead & 0x3FF) << 10) | (trail & 0x3FF));
+  out[0] = static_cast<char>(0xF0 | (codepoint >> 18));
+  out[1] = static_cast<char>(0x80 | ((codepoint >> 12) & 0x3F));
+  out[2] = static_cast<char>(0x80 | ((codepoint >> 6) & 0x3F));
+  out[3] = static_cast<char>(0x80 | (codepoint & 0x3F));
+}
+
+// Convert UTF-16 with potentially invalid surrogates to UTF-8.
+// Invalid surrogates are replaced with U+FFFD.
+void convertInvalidUtf16ToUtf8(const char16_t* input, size_t length, char* out) {
+  size_t position = 0;
+  bool pendingSurrogate = false;
+
+  for (size_t i = 0; i < length; i++) {
+    char16_t c = input[i];
+
+    if (pendingSurrogate) {
+      if (isTrailSurrogate(c)) {
+        encodeSurrogatePair(input[i - 1], c, out + position);
+        position += 4;
+        pendingSurrogate = false;
+      } else {
+        position += encodeUtf8CodeUnit(0xFFFD, out + position);
+        if (!isLeadSurrogate(c)) {
+          position += encodeUtf8CodeUnit(c, out + position);
+          pendingSurrogate = false;
+        }
+      }
+    } else if (isLeadSurrogate(c)) {
+      pendingSurrogate = true;
+    } else {
+      if (isTrailSurrogate(c)) {
+        position += encodeUtf8CodeUnit(0xFFFD, out + position);
+      } else {
+        position += encodeUtf8CodeUnit(c, out + position);
+      }
+    }
+  }
+
+  if (pendingSurrogate) {
+    encodeUtf8CodeUnit(0xFFFD, out + position);
+  }
+}
+
 }  // namespace
 
-jsg::BufferSource TextEncoder::encode(jsg::Lock& js, jsg::Optional<jsg::JsString> input) {
-  auto str = input.orDefault(js.str());
-  auto view = JSG_REQUIRE_NONNULL(jsg::BufferSource::tryAlloc(js, str.utf8Length(js)), RangeError,
-      "Cannot allocate space for TextEncoder.encode");
-  [[maybe_unused]] auto result = encodeIntoImpl(js, str, view);
-  KJ_DASSERT(result.written == view.size());
-  return kj::mv(view);
+jsg::Ref<TextEncoder> TextEncoder::constructor(jsg::Lock& js) {
+  return js.alloc<TextEncoder>();
+}
+
+jsg::JsUint8Array TextEncoder::encode(jsg::Lock& js, jsg::Optional<jsg::JsString> input) {
+  jsg::JsString str = input.orDefault(js.str());
+  std::shared_ptr<v8::BackingStore> backingStore;
+  size_t utf8_length = 0;
+
+  // Fast path: check if string is one-byte before creating ValueView
+  if (str.isOneByte(js)) {
+    auto length = str.length(js);
+    // Allocate buffer for Latin-1. Use v8::ArrayBuffer::NewBackingStore to avoid creating
+    // JS objects during conversion.
```

Comment:
FWIW, while this may not create JS objects it can and will trigger GC.. so let's try to do this (allocating the on-heap result buffer) once per call. All other allocations should be off-heap.